Mathlib/Algebra/MvPolynomial/Basic.lean

```diff
@@ -875,6 +875,43 @@ theorem C_dvd_iff_dvd_coeff (r : R) (φ : MvPolynomial σ R) : C r ∣ φ ↔ 
     IsRegular (∏ n in s, X n : MvPolynomial σ R) :=
   IsRegular.prod fun _ _ ↦ isRegular_X
 
+/-- The finset of nonzero coefficients of a multivariate polynomial. -/
+def coeffs (p : MvPolynomial σ R) : Finset R :=
+  letI := Classical.decEq R
+  Finset.image p.coeff p.support
+
+lemma coeffs_zero : coeffs (0 : MvPolynomial σ R) = ∅ :=
+  rfl
+
+lemma coeffs_one : coeffs (1 : MvPolynomial σ R) ⊆ {1} := by
+  classical
+    rw [coeffs, Finset.image_subset_iff]
+    simp_all [coeff_one]
+
+@[nontriviality]
+lemma coeffs_eq_empty_of_subsingleton [Subsingleton R] (p : MvPolynomial σ R) : p.coeffs = ∅ := by
+  simpa [coeffs] using Subsingleton.eq_zero p
+
+@[simp]
+lemma coeffs_one_of_nontrivial [Nontrivial R] : coeffs (1 : MvPolynomial σ R) = {1} := by
+  apply Finset.Subset.antisymm coeffs_one
+  simp only [coeffs, Finset.singleton_subset_iff, Finset.mem_image]
+  exact ⟨0, by simp⟩
+
+lemma mem_coeffs_iff {p : MvPolynomial σ R} {c : R} :
+    c ∈ p.coeffs ↔ ∃ n ∈ p.support, c = p.coeff n := by
+  simp [coeffs, eq_comm, (Finset.mem_image)]
+
+lemma coeff_mem_coeffs {p : MvPolynomial σ R} (m : σ →₀ ℕ)
+    (h : p.coeff m ≠ 0) : p.coeff m ∈ p.coeffs :=
+  letI := Classical.decEq R
+  Finset.mem_image_of_mem p.coeff (mem_support_iff.mpr h)
+
+lemma zero_not_mem_coeffs (p : MvPolynomial σ R) : 0 ∉ p.coeffs := by
+  intro hz
+  obtain ⟨n, hnsupp, hn⟩ := mem_coeffs_iff.mp hz
+  exact (mem_support_iff.mp hnsupp) hn.symm
+
```

Comment:
```suggestion
  simp [mem_coeffs_iff, mem_support_iff]
  tauto
```

Reply:
This seems to be a factor 10 slowdown in heartbeats. I would rather keep the explicit version.

Reply:
Oh sure!